### ISSUE TITLE
Question about Coross-Validation for a downstream task

### ISSUE THREAD

**Issue by PaulForInvent** [UKPLab/sentence-transformers#984]
Hey,

do you think, do you think I should use a cross-validation of my trainingdata while fine-tune a model for semantic search (and simalirity task)?

Surprisingly I always ignored this...

**Comment by nreimers** [comment]
If you perform an ablation on e.g. what is the best model, what is the best loss, what are the best parameters, then using CV can make sense if it is computationally feasible.

**Comment by PaulForInvent** [comment]
@nreimers Thanks.

Just try to do it, but I saw that for kfold you need of course the RandomSubsetSamplers. In my case I use the SentencesLabelDataset which is a IterableDataset and cannot be used with a sampler. That is bad.

Is it possible to have the SentencesLabelDataset  as normal Dataset?

**Comment by nreimers** [comment]
It would be better to first create the fold, and the re-init your SentencesLabelDataset

**Comment by PaulForInvent** [comment]
> It would be better to first create the fold, and the re-init your SentencesLabelDataset

So you suppose to create the fold without any pytorch dataset? But isn't it possible to chnage the SentencesLabelDataset to a normal dataset by replacing yield by return eg...?

**Comment by nreimers** [comment]
I think it is easier to first create your different folds, and then create a new SentencesLabelDataset  from it.

**Comment by PaulForInvent** [comment]
> I think it is easier to first create your different folds

For this I like to do it with a dataset and a SubsetSampler to sample the folds in a pytorch way? Or how would you create the fold?

**Comment by PhilipMay** [comment]
Maybe you want to have a look here: https://github.com/German-NLP-Group/xlsr

There I use cross validation as I think it is useful.

**Comment by PhilipMay** [comment]
I prefer to use crossvalidation when I do automated hyperparameter search. The reasons are:
- cross validation reduced overfitting on the validation set when you to automated hyperparameter search
- through using multiple val. sets they better cover your data space when working with small datasets
- because neural networks are random initialized the random effects on the results are reduced when you calculate the mean of the folds

**Comment by PaulForInvent** [comment]
@PhilipMay Thanks, Maybe using simple arrays is feasible also I wanted to do it like here:
https://www.machinecurve.com/index.php/2021/02/03/how-to-use-k-fold-cross-validation-with-pytorch/

But I think the SentencesLabelDataset  can be rewritten to a simple dataset.

**Comment by PhilipMay** [comment]
@PaulForInvent 

Here is the Optuna Importance Plot

![grafik](https://user-images.githubusercontent.com/229382/120807760-969b2800-c548-11eb-9d28-c34e8a88c447.png)

**Comment by PhilipMay** [comment]
@PaulForInvent 

and the slice plot

![grafik](https://user-images.githubusercontent.com/229382/120807848-b0d50600-c548-11eb-8ab7-4ca35f63c37c.png)

**Comment by PaulForInvent** [comment]
I asked this myself too.

https://github.com/UKPLab/sentence-transformers/issues/791

**Comment by PaulForInvent** [comment]
Now I have a different issue. Since I use mainly Batchhard-Losses, I have examples with its class labels. Up to now for evaluation I used a ranking metric on a different validation set. Now, I wonder how do I evaluate my model on each fold, as now, both data are similar structured (meaning these are just labeled examples).  Now I could use a evaluation metric if the class label is predicted correctly (multi class task) or a triplet Evaluator...


My main task is actually Ranking, so I also would like to do a ranking evaluation for each fold...but since my fold is fixed I cannot do a ranking task and just have to use the available samples of each class (possibly with a ParaphraseMiningEvaluator())?

**Comment by PaulForInvent** [comment]
@nreimers : 

I wonder if your ParaphraseMiningEvaluator or BinaryClassificationEvaluator handles the case for ignoring self refeernces by calculating the cosine score of a list of sentences with itself?

**Comment by nreimers** [comment]
It computes whatever you pass as your data. The ParaphraseMiningEvaluator ignores self references.

**Comment by PaulForInvent** [comment]
> The ParaphraseMiningEvaluator ignores self references.

Where exactly do you ignore them in the code?

For eg BinaryClassificationEvaluator I guess self references are not ignored if you pass for sentence2 same as sentence1?

**Comment by PaulForInvent** [comment]
@PhilipMay I just saw that your are drawing the parameters in each fold new. I did this same thing too. But shouldn't be the parameters for all folds the same?

**Comment by PhilipMay** [comment]
> I just saw that your are drawing the parameters in each fold new. 
No. It just seems like that. When you draw them from optuna multiple times the 2nd and alls following times it returns the same value until the trial is over.

> I did this same thing too. But shouldn't be the parameters for all folds the same?
It should (must) be all the same.

**Comment by PhilipMay** [comment]
> Should I then retrain the model using all trainingdata ? Also, despite setting a seed, you cannot guarantee that each model trained with the same parameters yield the same results... So you should save each model during CV and then continue finetuning on all the data?
Any standard way you experiences to be good? @nreimers

I hate seeds and do not use them when doing HP optimization with CV. I just do many CV steps and average them. CV is only about Hyperparameter finding and not about model creation.

When I want to create the "best" final model I train the model with same HP on the full dataset at the end.

**Comment by PaulForInvent** [comment]
> HP optimization with CV

Also my finding right now are that this is really difficult as the model really overfits during my k fold validation. Actually all models are very optimistic for any parameters. Maybe I should exclude explicitly some classes for each fold.

**Comment by PaulForInvent** [comment]
@PhilipMay What are your experience with randomness? If I do HP search and try to retrain the model with any parameters I get always different results. So, just finding the HPs is not meaningful as not reproducible...

**Comment by PhilipMay** [comment]
@PaulForInvent just because there is randomness and you get different results does not mean it is not useful.

For example:
I have usecases with small data sets (6000) where I do 10 fold x-validation. The result is the mean of the folds. That helps
to reduce the effect of randomness.

**Comment by PhilipMay** [comment]
By the way. I saw that the stsb dataset has duplicate sentences in the train set.
So doing cross validation might be no good idea since you might have information leakage from
train to validation...

@PaulForInvent
